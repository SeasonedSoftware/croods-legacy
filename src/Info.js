import { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import withOptions from './withOptions'
import mapStateToProps from './mapStateToProps'
import mapDispatchToProps from './mapDispatchToProps'
import setOrFetchInfo from './setOrFetchInfo'
import renderIfPresent from './renderIfPresent'
import renderInfoLoading from './renderInfoLoading'

class Info extends Component {
  constructor(props) {
    super(props)
    setOrFetchInfo(props)
  }

  componentDidUpdate(prevProps) {
    const { id } = this.props
    const { id: oldId } = prevProps

    if (id.toString() !== oldId.toString()) {
      setOrFetchInfo(this.props)
    }
  }

  render() {
    const { render, info, infoError, renderError } = this.props

    return (
      renderIfPresent(renderError, infoError) ||
      renderInfoLoading(this.props) ||
      render(info, this.props)
    )
  }
}

Info.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  info: PropTypes.object.isRequired,
  render: PropTypes.func.isRequired,
  parentId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  renderError: PropTypes.func,
}

// INTERNAL PROPS
// infoError: object,

export default withOptions(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Info),
)
