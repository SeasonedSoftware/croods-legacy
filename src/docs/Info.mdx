---
name: Info
route: /info
order: 9
---

# Info

The Info component provides a way to fetch the info about a resource and render component using the resource info.
This component does a `GET` request to the `path` and renders the render function.

## API

| Property      | Type          | Required | Default                  | Description                                                                                                                          |
| ------------- | ------------- | -------- | ------------------------ | ------------------------------------------------------------------------------------------------------------------------------------ |
| id            | String/Number | true     | -                        | Ex: 1234 or '1234'                                                                                                                   |
| name          | String        | true     | -                        | It must be equal to the reducer name set in your `rootReducer.js` and will be used as `path` if none is provided Ex: `baseUrl/name`. |
| render        | Func          | true     | -                        | A function returning a React Node. Ex: (info = {}, props) -> < JSX />                                                                |
| parentId      | String/Number | false    | -                        | Parent element's Id, for more specificity in changes.                                                                                |
| path          | String        | false    | `/${name}/${id}`         | Ex: /foo/123                                                                                                                         |
| renderError   | Func          | false    | Provider's renderError   | A function returning a React Node that will render when occurs an error. Ex: (error = {}) -> < JSX />                                |
| renderLoading | Func          | false    | Provider's renderLoading | A function returning a React Node that will be rendered while wait the server response. Ex: () -> < JSX />                           |
| disableCache  | Bool          | false    | false                    | Disable chache. When true, will force Croods to make a new API request each time the component renders.                              |

## How to use

Don't forget to add a reducer with the same name used here in your rootReducer, with **createReducer**

```javascript
import { Info } from 'croods'

export default (
  <Info
    name="pets"
    id="123"
    // if path is different than /pets/${id}, you can set a custom path like so:
    // path='/customPath'
    render={info => <Component info={info} />}
    renderError={error => <div>{error}</div>}
    disableCache
  />
)

```

### render

The render function receives the parameter **info** containing the response body.

### renderError

The renderError function is fired when the server response has a 4xx or 5xx status.
It receives the `error` parameter containing the error message.

## Example

You can see a working sample of this component's usage [here](https://github.com/SeasonedSoftware/croods/blob/master/example/src/colors/Info.js).

