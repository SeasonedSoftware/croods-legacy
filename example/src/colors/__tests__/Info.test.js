import React from 'react'
import renderer from 'react-test-renderer'

import Info from '../Info'

jest.mock('croods', () => ({
  Info: props => (
    <div {...props}>
      Info - {props.render({ id: 1, name: 'color name', color: 'color hexadecimal' })}
    </div>
  ),
}))

jest.mock('react-router-dom/Link', () => props => <div {...props}>Link - {props.children}</div>)

it('renders correctly', () => {
  const props = {
    match: {
      params: {
        id: 1,
      },
    },
  }
  const tree = renderer.create(<Info {...props} />).toJSON()
  expect(tree).toMatchSnapshot()
})
