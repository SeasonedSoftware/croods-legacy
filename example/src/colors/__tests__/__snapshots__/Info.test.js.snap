// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id={1}
  name="colors"
  render={[Function]}
>
  Info - 
  <h1
    style={
      Object {
        "color": "color hexadecimal",
      }
    }
  >
    color name
  </h1>
  <h2>
    color hexadecimal
  </h2>
  <div
    to="/"
  >
    Link - 
    Back
  </div>
</div>
`;
